keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1285 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP extends Kubernetes to configure a job policy for handling pod failures.
+In particular, the extension allows determining some of pod failures as caused
+by infrastructure errors and to retry them without incrementing the counter
+towards `backoffLimit`.
+
+Additionally, the extension allows determining some pod failures as caused by
+software bugs and to terminate the associated job early. This is needed to save
+time and computational resources wasted due to unnecessary retries of containers
+destined to fail due to software bugs.
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Running a large computational workload, comprising thousands of pods on
+thousands of nodes requires usage of pod restart policies in order
+to account for infrastructure failures.
+
+Currently, kubernetes Job API offers a way to account for infrastructure
+failures by setting `.backoffLimit > 0`. However, this mechanism intructs the
+job controller to restart all failed pods - regardless of the root cause
+of the failures. Thus, in some scenarios this leads to unnecessary
+restarts of many pods, resulting in a waste of time and computational
+resources. What makes the restarts more expensive is the fact that the
+failures may be encountered late in the execution time of a program.
+
+Sometimes it can be determined from containers exit codes
```

Comment:
Is it possible to achieve this at pod Level, it seems reusable. FYI I'm going to add a newfield `maxRestartTimes` to pod spec https://github.com/kubernetes/enhancements/pull/3339, it's still under discussion and waiting for review. However, if the idea accepted, the feature is somehow overlap with `backoffLimit`. This situation is similar to what this KEP wants to introduce.

Reply:
I think your proposal will be a nice building block we can easily build on in the jobs.

Reply:
when the PodSpec's `restartPolicy=Never`, it doesn't make sense to have an exitCode policy in the PodSpec. So in that case we will still need a policy in the JobSpec (this proposal).

In other words, the proposals can coexist.